packages/block-editor/src/hooks/layout.js

```diff
@@ -354,76 +418,63 @@ export const withInspectorControls = createHigherOrderComponent(
-		const selector = `.wp-container-${ id }.wp-container-${ id }`;
-		const [ blockGapSupport ] = useSettings( 'spacing.blockGap' );
-		const hasBlockGapSupport = blockGapSupport !== null;
-
-		// Get CSS string for the current layout type.
-		// The CSS and `style` element is only output if it is not empty.
-		let css;
-		if ( shouldRenderLayoutStyles ) {
-			const fullLayoutType = getLayoutType(
-				usedLayout?.type || 'default'
-			);
-			css = fullLayoutType?.getLayoutStyle?.( {
-				blockName: name,
-				selector,
-				layout: usedLayout,
-				style: attributes?.style,
-				hasBlockGapSupport,
-			} );
+		if ( ! hasLayoutBlockSupport( props.name ) ) {
+			return <BlockListBlock { ...props } />;
 		}
 
-		// Attach a `wp-container-` id-based class name as well as a layout class name such as `is-layout-flex`.
-		const layoutClassNames = classnames(
-			{
-				[ `wp-container-${ id }` ]: shouldRenderLayoutStyles && !! css, // Only attach a container class if there is generated CSS to be attached.
-			},
-			layoutClasses
-		);
-
-		const { setStyleOverride, deleteStyleOverride } = unlock(
-			useDispatch( blockEditorStore )
-		);
-
-		useEffect( () => {
-			if ( ! css ) return;
-			setStyleOverride( selector, { css } );
-			return () => {
-				deleteStyleOverride( selector );
-			};
-		}, [ selector, css, setStyleOverride, deleteStyleOverride ] );
-
 		return (
-			<BlockListBlock
-				{ ...props }
-				__unstableLayoutClassNames={ layoutClassNames }
+			<BlockListWithLayoutStyles
+				block={ BlockListBlock }
+				props={ props }
 			/>
 		);
 	},
 	'withLayoutStyles'
 );
 
+function BlockListWithChildLayoutStyles( { block: BlockListBlock, props } ) {
+	const { style: { layout = {} } = {} } = props.attributes;
+	const { selfStretch, flexSize } = layout;
+	const disableLayoutStyles = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		return !! getSettings().disableLayoutStyles;
+	} );
+
+	const id = useInstanceId( BlockListBlock );
+	const selector = `.wp-container-content-${ id }`;
+
+	let css = '';
+	if ( selfStretch === 'fixed' && flexSize ) {
+		css += `${ selector } {
+				flex-basis: ${ flexSize };
+				box-sizing: border-box;
+			}`;
+	} else if ( selfStretch === 'fill' ) {
+		css += `${ selector } {
+				flex-grow: 1;
+			}`;
+	}
+
+	// Attach a `wp-container-content` id-based classname.
+	const className = classnames( props?.className, {
+		[ `wp-container-content-${ id }` ]: ! disableLayoutStyles && !! css, // Only attach a container class if there is generated CSS to be attached.
+	} );
+
+	const { setStyleOverride, deleteStyleOverride } = unlock(
+		useDispatch( blockEditorStore )
+	);
+
+	useEffect( () => {
+		if ( ! css ) return;
+		setStyleOverride( selector, { css } );
+		return () => {
+			deleteStyleOverride( selector );
+		};
+	}, [ selector, css, setStyleOverride, deleteStyleOverride ] );
+
+	return <BlockListBlock { ...props } className={ className } />;
```

Comment:
There's one thing about the child layout code that is buggy, or at least inefficient: if `disableLayoutStyles` is true, and if the block has a child layout (i.e., there are valid `flexSize` or `selfStretch` values), the corresponding CSS will be created and added to style overrides. But the element _won't_ get the `wp-container-content-n` CSS class, and therefore the generated CSS won't be applied to any element, it's redundant.

The `withLayoutStyles` HOC is more efficient: it's doesn't even generate the CSS if layout styles are disabled. It would be nice if `withChildLayoutStyles` behaved the same way.

Cc @tellthemachines to verify that the argument above is correct and that there's no hidden gotcha.

Reply:
Oh that sounds like a bug, yes. The child layout CSS shouldn't be generated at all if `disableLayoutStyles` is true.